vault/identity_store_oidc_provider.go

```diff
@@ -1563,6 +1579,33 @@ func (i *IdentityStore) pathOIDCAuthorize(ctx context.Context, req *logical.Requ
 		scopes:      scopes,
 	}
 
+	// Validate the optional Proof Key for Code Exchange (PKCE) if a code
+	// challenge and code challenge method are provided. See details at
+	// https://datatracker.ietf.org/doc/html/rfc7636.
+	if codeChallengeRaw, ok := d.GetOk("code_challenge"); ok {
+		codeChallenge := codeChallengeRaw.(string)
+
+		// Validate the code challenge method
+		codeChallengeMethod := d.Get("code_challenge_method").(string)
+		switch codeChallengeMethod {
+		case codeChallengeMethodPlain, codeChallengeMethodS256:
+		case "":
+			codeChallengeMethod = codeChallengeMethodPlain
+		default:
+			return authResponse("", state, ErrAuthInvalidRequest, "invalid code_challenge_method")
+		}
+
+		// Validate the code challenge
+		if codeChallenge == "" {
```

Comment:
Wondering if we should enforce the minimum length of 43 characters on the server.

Reply:
> ... minimum length of 43 characters and a maximum length of 128 characters.

Since this is defined by the spec, I think we should enforce both the min and the max.

Reply:
Ah, I just realized that those min/max values are for the client's code_verifier. In this case, we're taking in the code_challenge. I wonder if we should still enforce some length bounds here 🤔 

We could still enforce that the code_challenge length is between [43, 128]. For the plain method, code_challenge = code_verifier, so that works out nicely. For the S256 method, code_challenge _should_ have length 43 if it was computed correctly by the client, since sha256(code_verifier) -> 32 bytes and base64url(32 bytes) -> 43 characters.

Reply:
Done in [b93eb5a](https://github.com/hashicorp/vault/pull/13917/commits/b93eb5a4af96e689090a022537ed19a48cf6097b). Added some tests as well.